### ISSUE TITLE
Crash: "failed to calculate pool rewards" in 2 node setup on both nodes

### ISSUE THREAD

**Issue by HM4704** [iotaledger/iota-core#776]
The following call stack is identical on both nodes:

```
panic: failed to calculate pool rewards for account 0x907c02e9302e0f0571f10f885594e56d8c54ff0708ab7a39bc1b74d396b93b12: failed to calculate pool reward without fixed costs due to overflow for slot 8196: 2523412566627373056 * 9: integer under- or overflow

goroutine 218 [running]:
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance.(*Tracker).ApplyEpoch.func1({0x90, 0x7c, 0x2, 0xe9, 0x30, 0x2e, 0xf, 0x5, 0x71, 0xf1, ...}, ...)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go:266 +0x488
github.com/iotaledger/hive.go/ds/shrinkingmap.(*ShrinkingMap[...]).ForEach(0x2ca0c40, 0x4015a555a0)
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/ds@v0.0.0-20240216135101-261e99d9d84a/shrinkingmap/shrinkingmap.go:174 +0x258
github.com/iotaledger/iota-core/pkg/core/account.(*Accounts).ForEach(...)
	/home/pi/Projects/crypto/iota/iota-core/pkg/core/account/accounts.go:118
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance.(*Tracker).ApplyEpoch(0x40003e5c20, 0x0, 0x40147e4228)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go:224 +0x270
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1.(*SybilProtection).CommitSlot(0x400034ac60, 0x2004)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:189 +0x2a0
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).createCommitment(0x40004281e0, 0x2004)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:230 +0x478
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).tryCommitSlotUntil(0x40004281e0, 0x2007)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:195 +0x8c
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).tryCommitUntil(0x40004281e0, 0x2007)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:120 +0x54
github.com/iotaledger/iota-core/components/protocol.provide.func1.NewProvider.14.1.1(0x40022b5080)
	/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:78 +0x118
github.com/iotaledger/hive.go/runtime/event.(*Event1[...]).Trigger.func1.1()
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/event/events.go:86 +0x2c
github.com/iotaledger/hive.go/runtime/workerpool.(*Task).run(0x4018e79a40)
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/task.go:48 +0x74
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).workerReadLoop(0x400028a690)
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:226 +0x28
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).worker(0x400028a690)
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:206 +0x54
created by github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).startWorkers in goroutine 201
	/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:198 +0x2c
```

**Comment by alexsporn** [comment]
@HM4704 which setup did you use? Did you modify any of the protocol values for your network?

**Comment by HM4704** [comment]
I am using the setup as described in [https://github.com/HM4704/iota-core-howto/blob/main/setup-2node-network.md.](https://github.com/HM4704/iota-core-howto/blob/main/setup-2node-network.md.)

The error happened once after about 24h runtime for the validator.

What protocol values would be of interest?

**Comment by HM4704** [comment]
The error happened again now (again on both nodes). It seems its always happening  after around 8194 slots have elapsed.

```
panic: failed to calculate pool rewards for account 0x907c02e9302e0f0571f10f885594e56d8c54ff0708ab7a39bc1b74d396b93b12: failed to calculate pool reward without fixed costs due to overflow for slot 16388: 2521075650825236480 \* 9: integer under- or overflow

goroutine 214 \[running\]:  
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance.(\*Tracker).ApplyEpoch.func1({0x90, 0x7c, 0x2, 0xe9, 0x30, 0x2e, 0xf, 0x5, 0x71, 0xf1, ...}, ...)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go:266 +0x488  
github.com/iotaledger/hive.go/ds/shrinkingmap.(\*ShrinkingMap\[...\]).ForEach(0x2ca0c40, 0x40007af5a0)  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/ds@v0.0.0-20240216135101-261e99d9d84a/shrinkingmap/shrinkingmap.go:174 +0x258  
github.com/iotaledger/iota-core/pkg/core/account.(\*Accounts).ForEach(...)  
/home/pi/Projects/crypto/iota/iota-core/pkg/core/account/accounts.go:118  
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance.(\*Tracker).ApplyEpoch(0x400007fd40, 0x1, 0x4027186ee8)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go:224 +0x270  
github.com/iotaledger/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1.(\*SybilProtection).CommitSlot(0x4000490c60, 0x4004)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:189 +0x2a0  
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(\*Manager).createCommitment(0x40000fe1e0, 0x4004)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:230 +0x478  
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(\*Manager).tryCommitSlotUntil(0x40000fe1e0, 0x4007)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:195 +0x8c  
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(\*Manager).tryCommitUntil(0x40000fe1e0, 0x4007)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:120 +0x54  
github.com/iotaledger/iota-core/components/protocol.provide.func1.NewProvider.14.1.1(0x40030f8900)  
/home/pi/Projects/crypto/iota/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:78 +0x118  
github.com/iotaledger/hive.go/runtime/event.(\*Event1\[...\]).Trigger.func1.1()  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/event/events.go:86 +0x2c  
github.com/iotaledger/hive.go/runtime/workerpool.(\*Task).run(0x4026a5bb00)  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/task.go:48 +0x74  
github.com/iotaledger/hive.go/runtime/workerpool.(\*WorkerPool).workerReadLoop(0x40002a05b0)  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:226 +0x28  
github.com/iotaledger/hive.go/runtime/workerpool.(\*WorkerPool).worker(0x40002a05b0)  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:206 +0x54  
created by github.com/iotaledger/hive.go/runtime/workerpool.(\*WorkerPool).startWorkers in goroutine 192  
/home/pi/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240216135101-261e99d9d84a/workerpool/workerpool.go:198 +0x2c
```

**Comment by muXxer** [comment]
Thanks for the report. The main difference is that you have less validators than we have I guess. Ofc it shouldn't panic in that case. We will investigate :+1:

**Comment by HM4704** [comment]
I have looked into the code and I hope I understand now better what is happening. The failing function is only called for each epoch which is 2^13(8192) Slots.

I have added a log output for the values that are multiplied:

INFO    Protocol.Engine0.PerformanceTracker  targetReward: 614926977410990   poolCoefficient: 4096

**Comment by PhilippGackstatter** [comment]
I think our `deriveTargetRewardsRates` might be wrong. According to this comment in `poolReward`:

```
// Since `Pool Coefficient` uses at most 12 bits, `Target Reward(n)` uses at most 41 bits, and `Performance Factor` uses at most 8 bits,
// this multiplication will not overflow using uint64 variables.
```

the `targetReward` should use at most 41 bits. But adding this check in `deriveTargetRewardsRates`:

```
initialRewardBits := math.Log2(initialTargetRewardsRate)
if initialRewardBits > 41.0 {
	panic(fmt.Sprintf("initialTargetRewardsRate must use at most 41 bits, but uses %d", uint8(initialRewardBits)))
}
```

panics with

```
panic: initialTargetRewardsRate must use at most 41 bits, but uses 49 bits
```

@oliviasaa can you double-check?

If that's the case, then we should consider adding this as a "sanity check".

**Comment by oliviasaa** [comment]
I'm checking again all the bits math again. Something went obviously wrong at some point; I'll push a fix during the day. 

@PhilippGackstatter, yes, I'll also add another sanity check to this.

**Comment by PhilippGackstatter** [comment]
I was able to reproduce this issue within 1 minute by reducing the slot duration to 4 seconds and the number of slots in an epoch to 16, but, importantly, keeping the `InitialTargetRewardsRate` the same as if the epochs were 2^13 slots long (since the rewards rate is then simply larger).

In order to reproduce we have to change the docker protocol params:

```
ProtocolParamsDocker = func() *iotago.V3ProtocolParameters {
	params := iotago.NewV3SnapshotProtocolParameters(
		iotago.WithNetworkOptions(fmt.Sprintf("docker-%d", time.Now().Unix()), iotago.PrefixTestnet),
		iotago.WithTimeProviderOptions(5, time.Now().Unix(), 4, 4),
		iotago.WithLivenessOptions(8, 10, 3, 5, 6),
		iotago.WithCongestionControlOptions(1, 0, 0, 400_000, 300_000, 100_000, 1000, 100),
	)

	// Set the values we would use for our regular number of slots in an epoch (2^13 slots).
	params.RewardsParameters().InitialTargetRewardsRate = 616067521149261
	params.RewardsParameters().FinalTargetRewardsRate = 226702563632670

	return params
}()
```

After running this for a minute (1 epoch) we get almost the exact same error as in the original issue:

```
panic: failed to calculate pool rewards for account 0x907c02e9302e0f0571f10f885594e56d8c54ff0708ab7a39bc1b74d396b93b12: failed to calculate pool reward without fixed costs due to overflow for slot 20: 2523412566627373056 * 8: integer under- or overflow
```

With that setup I ran the docker network again with [PR 786](https://github.com/iotaledger/iota-core/pull/786) (commit `c774d5fa661a27bdbf9f828768bd9f7441388a22`) and it did not panic.

**Comment by PhilippGackstatter** [comment]
@HM4704 This issue should be resolved now. If you want to test it, run iota-core with one of the latest commits and feel free to report back to us if it also works with epoch lengths of 8192 slots.

Thanks for reporting this issue!
